File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/NativeQueries/Mocks/MockMethodRef.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

#if !SILVERLIGHT
using System;
using System.Reflection;
using Db4oUnit;
using Db4objects.Db4o.Instrumentation.Api;
using Db4objects.Db4o.Tests.NativeQueries.Mocks;

namespace Db4objects.Db4o.Tests.NativeQueries.Mocks
{
	public class MockMethodRef : IMethodRef
	{
		private readonly MethodInfo _method;

		public MockMethodRef(MethodInfo method)
		{
			_method = method;
		}

		public virtual string Name
		{
			get
			{
				return _method.Name;
			}
		}

		public virtual ITypeRef[] ParamTypes
		{
			get
			{
				Type[] paramTypes = Sharpen.Runtime.GetParameterTypes(_method);
				ITypeRef[] types = new ITypeRef[paramTypes.Length];
				for (int i = 0; i < paramTypes.Length; ++i)
				{
					types[i] = TypeRef(paramTypes[i]);
				}
				return types;
			}
		}

		public virtual ITypeRef DeclaringType
		{
			get
			{
				return TypeRef(_method.DeclaringType);
			}
		}

		private ITypeRef TypeRef(Type type)
		{
			return new MockTypeRef(type);
		}

		public virtual ITypeRef ReturnType
		{
			get
			{
				return TypeRef(_method.ReturnType);
			}
		}

		public override string ToString()
		{
			return Name;
		}

		public override bool Equals(object obj)
		{
			if (!(obj is IMethodRef))
			{
				return false;
			}
			IMethodRef other = (IMethodRef)obj;
			return Name.Equals(other.Name) && Check.ObjectsAreEqual(DeclaringType, other.DeclaringType
				) && Check.ObjectsAreEqual(ReturnType, other.ReturnType) && Check.ArraysAreEqual
				(ParamTypes, other.ParamTypes);
		}
	}
}
#endif // !SILVERLIGHT

File: db4onqopt/test/src/com/db4o/test/nativequery/mocks/MockMethodRef.java
/* Copyright (C) 2004 - 2006 Versant Inc. http://www.db4o.com */

package com.db4o.test.nativequery.mocks;

import java.lang.reflect.*;

import com.db4o.instrumentation.api.*;

import db4ounit.*;

public class MockMethodRef implements MethodRef {

	private final Method _method;

	public MockMethodRef(Method method) {
		_method = method;
	}

	public String name() {
		return _method.getName();
	}

	public TypeRef[] paramTypes() {
		final Class[] paramTypes = _method.getParameterTypes();
		final TypeRef[] types = new TypeRef[paramTypes.length];
		for (int i=0; i<paramTypes.length; ++i) {
			types[i] = typeRef(paramTypes[i]);
		}
		return types;
	}
	
	public TypeRef declaringType() {
		return typeRef(_method.getDeclaringClass());
	}

	private TypeRef typeRef(Class type) {
		return new MockTypeRef(type);
	}

	public TypeRef returnType() {
		return typeRef(_method.getReturnType());
	}
	
	public String toString() {
		return name();
	}
	
	public boolean equals(Object obj) {
		if (!(obj instanceof MethodRef)) {
			return false;
		}
		
		MethodRef other = (MethodRef)obj;
		return name().equals(other.name())
			&& Check.objectsAreEqual(declaringType(), other.declaringType())
			&& Check.objectsAreEqual(returnType(), other.returnType())
			&& Check.arraysAreEqual(paramTypes(), other.paramTypes());
	}	
}
